<a href="https://colab.research.google.com/github/WaleedSharkas/hackerfile/blob/master/Exploring_Ebay_Car_Sales_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving autos.csv to autos.csv


# Introduction 

In this project, we will work with a dataset on used cars from eBay Kleinanzeigen, a classifieds section from the Germany eBay website. Although the full dataset consisted of over 370,000 listings, we will only use 50,000 data points. This will allow us to demonstrate how to effectively clean a data set from scratch.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken 
from this date.
-name - Name of the car.
-seller - Whether the seller is private or a dealer.
-offerType - The type of listing
-price - The price on the ad to sell the car.
-abtest - Whether the listing is included in an A/B test.
-vehicleType - The vehicle Type.
-yearOfRegistration - The year in which which year the car was first registered.
-gearbox - The transmission type.
-powerPS - The power of the car in PS.
-model - The car model name.
-kilometer - How many kilometers the car has driven.
-monthOfRegistration - The month in which which year the car was first registered.
-fuelType - What type of fuel the car uses.
-brand - The brand of the car.
-notRepairedDamage - If the car has a damage which is not yet repaired.
-dateCreated - The date on which the eBay listing was created.
-nrOfPictures - The number of pictures in the ad.
-postalCode - The postal code for the location of the vehicle.
-lastSeenOnline - When the crawler saw this ad last online.

The goal of this project is to clean the data and then use pandas to perform some basic initial analysis of the listings.

In [2]:
import pandas as pd
import numpy as np

In [3]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

Observations on current data:
- Contains 20 columns, most stored as strings
- Some Columns contain null values
- dateCrawled, LastSeen, price, odometer are stored as strings. Will need to be changed 
- certain coloumns will need to be re-named and converted from camelcase to snakecase


In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns= ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()

date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

We cleaned up the columns to make them less confusing and more clean:

-yearOfRegistration to registration_year

-monthOfRegistration to registration_month

-notRepairedDamage to unrepaired_damage

-dateCreated to ad_created

Also changed the columns from camelcase to snakecase due to it being the preferred style for Python.

In [6]:
autos.describe(include='all')

date_crawled         name  ...   postal_code            last_seen
count                 50000        50000  ...  50000.000000                50000
unique                48213        38754  ...           NaN                39481
top     2016-03-30 19:48:02  Ford_Fiesta  ...           NaN  2016-04-07 06:17:27
freq                      3           78  ...           NaN                    8
mean                    NaN          NaN  ...  50813.627300                  NaN
std                     NaN          NaN  ...  25779.747957                  NaN
min                     NaN          NaN  ...   1067.000000                  NaN
25%                     NaN          NaN  ...  30451.000000                  NaN
50%                     NaN          NaN  ...  49577.000000                  NaN
75%                     NaN          NaN  ...  71540.000000                  NaN
max                     NaN          NaN  ...  99998.000000                  NaN

[11 rows x 20 columns]

In [7]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [9]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [10]:
pd.to_datetime(autos["date_crawled"]).max()
# most recent date an observation was collected.

Timestamp('2016-04-07 14:36:56')

In [11]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].count()
# number of cars with registration years prior to 1906 or after 2016

1972

In [12]:
 # the number of cars with power rating less than 1ps or greater than 2000ps
autos.loc[~autos["power_ps"].between(1, 2000), "power_ps"].count()

5518

In [13]:
autos["registration_month"].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

Observations:
- we notice that the seller and offer_type columns are almost exclusively a single value. That means they are good candidates to drop as they won't have useful information for analysis. 
-It also appears that the column nr_of_pictures is all zeros, since the minimum and maximum values in the column are both zero, good candidates to drop as they won't have useful information for analysis. 
- the power_ps column has suspicious amount below 1ps
- spicious minimum value of 0, due to There being no zero month. Could be due to typos

In [14]:
autos = autos.drop(["num_photos", "seller", "offer_type"], axis=1)

In [15]:
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [16]:
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [17]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [18]:
print(autos["odometer_km"].nunique())

13


In [19]:
(autos["odometer_km"] >= 100000).sum()

39763

Overall, there doesnt seem to be anything that stands out in the odometer_km data.

In [20]:
autos["price"].nunique()

2357

In [21]:
print(autos["price"].describe())
autos["price"].value_counts().head(15)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64

In [22]:
autos["price"].value_counts().sort_index(ascending=True).head(30)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price, dtype: int64

In [23]:
autos["price"].value_counts().sort_index(ascending=False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

The price column has numbers in both the max and min list that do not seem unrealistic. Further breakdown of numbers is needed here.

In [24]:
autos.loc[autos["price"] > 350000, ["name", "model", "brand", "registration_year", "price"]]

name  ...     price
514                     Ford_Focus_Turnier_1.6_16V_Style  ...    999999
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000  ...  11111111
7814                                         Ferrari_F40  ...   1300000
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  ...  10000000
22947                         Bmw_530d_zum_ausschlachten  ...   1234566
24384                            Schlachte_Golf_3_gt_tdi  ...  11111111
27371                                         Fiat_Punto  ...  12345678
37585                                Volkswagen_Jetta_GT  ...    999990
39377                        Tausche_volvo_v40_gegen_van  ...  12345678
39705                        Tausch_gegen_gleichwertiges  ...  99999999
42221                                  Leasinguebernahme  ...  27322222
43049                                      2_VW_Busse_T3  ...    999999
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...  ...  12345678
47634                                        Ferrari_FXX  ...   3890000

[14 rows x 5 columns]

In [25]:
autos.loc[autos["price"] < 0, ["name", "model", "brand", "registration_year", "price"]]

Empty DataFrame
Columns: [name, model, brand, registration_year, price]
Index: []

In [26]:
autos = autos[autos["price"].between(100,1000000)]

Price Observations:
- There are 1,421 cars listed with $0 price, It is only 2% of data. We will wait to see if it needs to be removed. 
- It does seem odd that a ford focus would sell for that high. Not sure if it is worth confirming this information. 
- Allow for prices only in the range of 100 - 1,000,000 . Anything below 100 seems odd

In [27]:
autos[["date_crawled", "ad_created", "last_seen"]] = autos[["date_crawled", "ad_created", "last_seen"]].apply(pd.to_datetime)

In [28]:
autos[["date_crawled", "ad_created", "last_seen"]].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48227 entries, 0 to 49999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   date_crawled  48227 non-null  datetime64[ns]
 1   ad_created    48227 non-null  datetime64[ns]
 2   last_seen     48227 non-null  datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 1.5 MB


In [29]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled ad_created           last_seen
0 2016-03-26 17:47:46 2016-03-26 2016-04-06 06:45:54
1 2016-04-04 13:38:56 2016-04-04 2016-04-06 14:45:08
2 2016-03-26 18:57:24 2016-03-26 2016-04-06 20:15:37
3 2016-03-12 16:58:10 2016-03-12 2016-03-15 03:16:28
4 2016-04-01 14:38:50 2016-04-01 2016-04-01 14:38:50

In [30]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_index()

2016-03-05    0.025359
2016-03-06    0.014038
2016-03-07    0.036059
2016-03-08    0.033176
2016-03-09    0.033011
2016-03-10    0.032285
2016-03-11    0.032596
2016-03-12    0.036909
2016-03-13    0.015676
2016-03-14    0.036660
2016-03-15    0.034317
2016-03-16    0.029465
2016-03-17    0.031518
2016-03-18    0.012897
2016-03-19    0.034732
2016-03-20    0.037800
2016-03-21    0.037220
2016-03-22    0.032886
2016-03-23    0.032285
2016-03-24    0.029444
2016-03-25    0.031497
2016-03-26    0.032306
2016-03-27    0.031124
2016-03-28    0.034960
2016-03-29    0.034130
2016-03-30    0.033736
2016-03-31    0.031849
2016-04-01    0.033695
2016-04-02    0.035602
2016-04-03    0.038609
2016-04-04    0.036536
2016-04-05    0.013063
2016-04-06    0.003172
2016-04-07    0.001389
Name: date_crawled, dtype: float64

In [31]:
autos["date_crawled"].dt.floor("D").value_counts(normalize = True).sort_values()

2016-04-07    0.001389
2016-04-06    0.003172
2016-03-18    0.012897
2016-04-05    0.013063
2016-03-06    0.014038
2016-03-13    0.015676
2016-03-05    0.025359
2016-03-24    0.029444
2016-03-16    0.029465
2016-03-27    0.031124
2016-03-25    0.031497
2016-03-17    0.031518
2016-03-31    0.031849
2016-03-10    0.032285
2016-03-23    0.032285
2016-03-26    0.032306
2016-03-11    0.032596
2016-03-22    0.032886
2016-03-09    0.033011
2016-03-08    0.033176
2016-04-01    0.033695
2016-03-30    0.033736
2016-03-29    0.034130
2016-03-15    0.034317
2016-03-19    0.034732
2016-03-28    0.034960
2016-04-02    0.035602
2016-03-07    0.036059
2016-04-04    0.036536
2016-03-14    0.036660
2016-03-12    0.036909
2016-03-21    0.037220
2016-03-20    0.037800
2016-04-03    0.038609
Name: date_crawled, dtype: float64

In [32]:
autos["date_crawled"].dt.hour.value_counts(normalize = True).sort_index()

0     0.026914
1     0.015157
2     0.006552
3     0.005018
4     0.004147
5     0.003048
6     0.003442
7     0.014038
8     0.019491
9     0.035395
10    0.047753
11    0.055197
12    0.058660
13    0.065109
14    0.068675
15    0.066228
16    0.069961
17    0.070396
18    0.073859
19    0.067887
20    0.068115
21    0.061625
22    0.050822
23    0.042507
Name: date_crawled, dtype: float64

Data Crawled Observation

This data was crawled over daily and around once a month, from March 2016 to april 2016. The overall distribution is faily uniformly scrpaed over. Hourly breakdown shows non uniformly between 2am-6am. 

In [33]:
(autos["last_seen"].dt.floor("D").value_counts(normalize=True).sort_index())

2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005433
2016-03-08    0.007320
2016-03-09    0.009580
2016-03-10    0.010637
2016-03-11    0.012400
2016-03-12    0.023783
2016-03-13    0.008875
2016-03-14    0.012628
2016-03-15    0.015862
2016-03-16    0.016443
2016-03-17    0.028096
2016-03-18    0.007320
2016-03-19    0.015759
2016-03-20    0.020652
2016-03-21    0.020549
2016-03-22    0.021357
2016-03-23    0.018579
2016-03-24    0.019761
2016-03-25    0.019097
2016-03-26    0.016671
2016-03-27    0.015551
2016-03-28    0.020860
2016-03-29    0.022311
2016-03-30    0.024696
2016-03-31    0.023825
2016-04-01    0.022850
2016-04-02    0.024882
2016-04-03    0.025131
2016-04-04    0.024530
2016-04-05    0.125054
2016-04-06    0.221971
2016-04-07    0.132146
Name: last_seen, dtype: float64

Last Seen Observation

This data was crawled over daily and around once a month, from March 2016 to april 2016. Over half of the entries were last seen in the last five days of data.

In [34]:
autos["ad_created"].dt.floor("D").value_counts(normalize = True).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038858
2016-04-04    0.036888
2016-04-05    0.011798
2016-04-06    0.003255
2016-04-07    0.001244
Name: ad_created, Length: 76, dtype: float64

In [35]:
autos["ad_created"].dt.year.value_counts(normalize = True)

2016    0.999876
2015    0.000124
Name: ad_created, dtype: float64

In [36]:
autos["ad_created"].dt.hour.value_counts(normalize = True).sort_index()

0    1.0
Name: ad_created, dtype: float64

AD Created Observation
 
The ads were created between june 2015  and april 2016. 

In [37]:
autos["registration_year"].describe()

count    48227.000000
mean      2004.730151
std         87.894768
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Registratoin Year Observations

There are odd numbers here, it is not possible to have a car in the "1000" which is the min value. The max value is also odd with "9999". Any vehivle registred after 2016 should not be in the data. We should focus on 1900-2016. Further Breakdown needed.

In [38]:
autos.loc[~autos["registration_year"].between(1906, 2016), "registration_year"].value_counts()

2017    1383
2018     468
9999       3
5000       3
2019       2
1800       2
2800       1
6200       1
4500       1
8888       1
4800       1
1000       1
9000       1
1001       1
1111       1
5911       1
4100       1
Name: registration_year, dtype: int64

In [39]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]
# Percent of invalid data 

0.03881643063014494

In [40]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].describe()

count    46355.000000
mean      2002.939057
std          7.128186
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [41]:
autos.loc[autos["registration_year"].between(1995, 2016), "registration_year"].value_counts(normalize = True).sort_index()

1995    0.027747
1996    0.031515
1997    0.044773
1998    0.054425
1999    0.066961
2000    0.072194
2001    0.061147
2002    0.057611
2003    0.062635
2004    0.062775
2005    0.067706
2006    0.062054
2007    0.052867
2008    0.051401
2009    0.048401
2010    0.036911
2011    0.037632
2012    0.030422
2013    0.018630
2014    0.015397
2015    0.008838
2016    0.027957
Name: registration_year, dtype: float64

In [42]:
autos.loc[autos["registration_year"].between(1906, 2016), "registration_year"].value_counts(normalize = True).sort_index().loc[1970:].sum()

0.9955991802394564

Registratoin Year Observations Final

Due to around 99% of this data occuring after 1970. We will set a date range for this data " 1970 - 2016".

In [43]:
autos = autos[autos["registration_year"].between(1969,2016)]
autos["registration_year"].value_counts(normalize=True).head(30)

2000    0.067230
2005    0.063050
1999    0.062357
2004    0.058458
2003    0.058328
2006    0.057786
2001    0.056942
2002    0.053650
1998    0.050682
2007    0.049231
2008    0.047867
2009    0.045073
1997    0.041694
2011    0.035044
2010    0.034373
1996    0.029348
2012    0.028330
2016    0.026034
1995    0.025839
2013    0.017349
2014    0.014338
1994    0.013559
1993    0.009097
2015    0.008230
1992    0.007949
1991    0.007321
1990    0.007191
1989    0.003704
1988    0.002881
1985    0.002036
Name: registration_year, dtype: float64

In [44]:
autos["brand"].nunique()


40

In [45]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211977
bmw               0.110505
opel              0.107472
mercedes_benz     0.096556
audi              0.087091
ford              0.069266
renault           0.047195
peugeot           0.029955
fiat              0.025558
seat              0.018324
skoda             0.016439
nissan            0.015400
mazda             0.015291
smart             0.014252
citroen           0.014057
toyota            0.012844
hyundai           0.010050
volvo             0.009032
mini              0.008837
sonstige_autos    0.008642
mitsubishi        0.008209
honda             0.007906
kia               0.007104
alfa_romeo        0.006649
suzuki            0.005956
porsche           0.005913
chevrolet         0.005545
chrysler          0.003487
dacia             0.002664
daihatsu          0.002512
jeep              0.002296
subaru            0.002123
land_rover        0.002079
saab              0.001668
daewoo            0.001494
jaguar            0.001494
rover             0.001343
t

In [46]:
autos["brand"].value_counts(normalize = True).head(10).sum()

0.803898635477583

Brand Observations
- Volkswagen is the most popular brand, with aprox double bmw who is second most popular
- WE will limit our analysis to brands in the top 10 ">0.018"

In [47]:
brand_count = autos["brand"].value_counts(normalize=True)
common_brands = brand_count[brand_count > .018].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [48]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9375,
 'bmw': 8383,
 'fiat': 2804,
 'ford': 3846,
 'mercedes_benz': 8565,
 'opel': 3002,
 'peugeot': 3108,
 'renault': 2491,
 'seat': 4433,
 'volkswagen': 5629}

In [49]:
autos.groupby("brand")[["odometer_km", "price"]].mean().loc[common_brands].sort_values("price", ascending = False).round(0)

odometer_km   price
audi              129240.0  9376.0
mercedes_benz     131243.0  8566.0
bmw               132790.0  8384.0
volkswagen        128839.0  5629.0
seat              121537.0  4433.0
ford              124980.0  3846.0
peugeot           127184.0  3108.0
opel              129465.0  3003.0
fiat              117237.0  2805.0
renault           128424.0  2491.0

In [50]:
autos.groupby("brand")["model"].value_counts(dropna = False).sort_values(ascending = False).head(10)

brand          model   
volkswagen     golf        3684
bmw            3er         2602
volkswagen     polo        1592
opel           corsa       1567
volkswagen     passat      1345
opel           astra       1337
audi           a4          1226
mercedes_benz  c_klasse    1135
bmw            5er         1123
mercedes_benz  e_klasse     953
Name: model, dtype: int64

In [51]:
autos.groupby("odometer_km")["brand"].describe().sort_index().round(1)

count unique         top  freq
odometer_km                                
5000           681     38  volkswagen   141
10000          229     25  volkswagen    47
20000          731     35  volkswagen   135
30000          752     33  volkswagen   164
40000          790     34  volkswagen   143
50000          979     38  volkswagen   183
60000         1111     39  volkswagen   206
70000         1171     39  volkswagen   206
80000         1362     38  volkswagen   249
90000         1660     39  volkswagen   287
100000        2024     39  volkswagen   364
125000        4829     40  volkswagen   838
150000       29851     40  volkswagen  6824

In [52]:
autos.groupby("brand")["odometer_km"].describe().sort_index().round(1)

count      mean      std  ...       50%       75%       max
brand                                      ...                              
alfa_romeo       307.0  132573.3  32745.7  ...  150000.0  150000.0  150000.0
audi            4021.0  129240.2  38747.2  ...  150000.0  150000.0  150000.0
bmw             5102.0  132790.1  34493.3  ...  150000.0  150000.0  150000.0
chevrolet        256.0   99882.8  45460.5  ...  100000.0  150000.0  150000.0
chrysler         161.0  133882.0  33976.0  ...  150000.0  150000.0  150000.0
citroen          649.0  119537.8  40190.7  ...  150000.0  150000.0  150000.0
dacia            123.0   84268.3  48085.0  ...   80000.0  125000.0  150000.0
daewoo            69.0  121231.9  37957.1  ...  150000.0  150000.0  150000.0
daihatsu         116.0  116120.7  42616.0  ...  125000.0  150000.0  150000.0
fiat            1180.0  117237.3  42417.8  ...  150000.0  150000.0  150000.0
ford            3198.0  124979.7  39124.7  ...  150000.0  150000.0  150000.0
honda            365.0  122493.2  40206.6  ...  150000.0  150000.0  150000.0
hyundai          464.0  106885.8  44894.3  ...  125000.0  150000.0  150000.0
jaguar            69.0  125724.6  38139.4  ...  150000.0  150000.0  150000.0
jeep             106.0  127122.6  38612.5  ...  150000.0  150000.0  150000.0
kia              328.0  112301.8  45192.0  ...  125000.0  150000.0  150000.0
lada              27.0   83518.5  38774.9  ...   80000.0  100000.0  150000.0
lancia            49.0  122346.9  40454.5  ...  150000.0  150000.0  150000.0
land_rover        96.0  118489.6  46843.1  ...  150000.0  150000.0  150000.0
mazda            706.0  124553.8  39867.4  ...  150000.0  150000.0  150000.0
mercedes_benz   4458.0  131242.7  35755.5  ...  150000.0  150000.0  150000.0
mini             408.0   88308.8  43710.9  ...   90000.0  125000.0  150000.0
mitsubishi       379.0  127097.6  38261.1  ...  150000.0  150000.0  150000.0
nissan           711.0  118326.3  41675.4  ...  150000.0  150000.0  150000.0
opel            4962.0  129464.9  36924.7  ...  150000.0  150000.0  150000.0
peugeot         1383.0  127183.7  35962.1  ...  150000.0  150000.0  150000.0
porsche          273.0   99029.3  48196.1  ...  100000.0  150000.0  150000.0
renault         2179.0  128423.6  36638.5  ...  150000.0  150000.0  150000.0
rover             62.0  137661.3  26699.9  ...  150000.0  150000.0  150000.0
saab              77.0  144415.6  22197.9  ...  150000.0  150000.0  150000.0
seat             846.0  121536.6  40847.6  ...  150000.0  150000.0  150000.0
skoda            759.0  111014.5  45710.7  ...  125000.0  150000.0  150000.0
smart            658.0   99734.0  41866.3  ...  100000.0  150000.0  150000.0
sonstige_autos   399.0   92042.6  55285.0  ...  100000.0  150000.0  150000.0
subaru            98.0  125612.2  39029.9  ...  150000.0  150000.0  150000.0
suzuki           275.0  108163.6  43020.0  ...  125000.0  150000.0  150000.0
toyota           593.0  115944.4  41846.5  ...  125000.0  150000.0  150000.0
trabant           54.0   54722.2  36709.2  ...   60000.0   70000.0  150000.0
volkswagen      9787.0  128838.8  38334.1  ...  150000.0  150000.0  150000.0
volvo            417.0  139244.6  27795.1  ...  150000.0  150000.0  150000.0

[40 rows x 8 columns]

In [53]:
autos.groupby("odometer_km")["price"].describe().sort_index().round(1)

count     mean      std  ...      50%      75%       max
odometer_km                             ...                            
5000           681.0   9563.6  27988.3  ...   2000.0   7950.0  350000.0
10000          229.0  21420.5  21150.8  ...  16900.0  27777.0  163991.0
20000          731.0  18430.8  23273.9  ...  14400.0  24200.0  345000.0
30000          752.0  16731.9  16908.3  ...  12830.0  21922.5  220000.0
40000          790.0  15468.5  13659.5  ...  12422.5  19995.0  169999.0
50000          979.0  13721.1  12219.5  ...  10500.0  17970.0  180000.0
60000         1111.0  12352.2  12875.5  ...   9200.0  16000.0  190000.0
70000         1171.0  10893.8   9226.9  ...   8250.0  15000.0   63999.0
80000         1362.0   9731.2   8855.9  ...   7499.0  13000.0  105000.0
90000         1660.0   8392.9   8247.5  ...   5990.0  11412.5   69997.0
100000        2024.0   8069.7  10567.4  ...   5600.0  10700.0  299000.0
125000        4829.0   6435.2  15966.0  ...   4000.0   8450.0  999999.0
150000       29851.0   3845.4   9429.6  ...   2200.0   4800.0  999999.0

[13 rows x 8 columns]

Brand Observations Final

-Mercedes is the most expensive brand, C_klasse top model

-Top 3 brands are Mercedes, audi, and bmw

-Due to Volkswagen having a price that is an avg of the top 10, this 
might be a reason why it is the most sold brand. Top sold model is polo

-The least expensive in the top 10 are fiat and renault

-We can see that the higher milage, the less the price

In [54]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5629
bmw                  8383
opel                 3002
mercedes_benz        8565
audi                 9375
ford                 3846
renault              2491
peugeot              3108
fiat                 2804
seat                 4433

In [55]:

brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [56]:

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132790
mercedes_benz        131242
opel                 129464
audi                 129240
volkswagen           128838
renault              128423
peugeot              127183
ford                 124979
seat                 121536
fiat                 117237

In [57]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132790        8383
mercedes_benz        131242        8565
opel                 129464        3002
audi                 129240        9375
volkswagen           128838        5629
renault              128423        2491
peugeot              127183        3108
ford                 124979        3846
seat                 121536        4433
fiat                 117237        2804

Price and Milage observation

- Overall the Mileage does not very to much between diffrent brands
- In this list, the more expensive cars have higher prices versus less expensive have much lower prices 

# Conclusion

Overall, we went through some of the steps needed to effectively clean the data set from scratch on the used cars from eBay Kleinanzeigen, a classifieds section from the Germany eBay website. 

Steps taken were as follows:

- Cleaning Column Names
- Simple exploration and Cleaning 
- Exploring certain columns 
- Narrowing down data in regards to price "100-1,000,000" 
- Dealing with Incorrect Registration Dates
- Exploring diffrent comparissons such as price, brand, odomoter 